core/ErrorQueue.cc

```diff
@@ -87,20 +83,41 @@ vector<unique_ptr<core::Error>> ErrorQueue::drainAllErrors() {
     return move(drainWithQueryResponses().first);
 }
 
-void ErrorQueue::flushErrors(bool all) {
+void ErrorQueue::flushAllErrors(const GlobalState &gs) {
     checkOwned();
     if (ignoreFlushes) {
         return;
     }
 
-    Timer timeit(tracer, "ErrorQueue::flushErrors");
-    vector<unique_ptr<ErrorQueueMessage>> errors;
-    if (all) {
-        errors = drainAll();
-    } else {
-        errors = drainFlushed();
+    Timer timeit(tracer, "ErrorQueue::flushAllErrors");
+    auto collectedErrors = drainAll();
+
+    for (auto &it : collectedErrors) {
+        errorFlusher->flushErrors(logger, move(it.second), gs, it.first);
     }
-    errorFlusher->flushErrors(logger, move(errors));
+}
+
+void ErrorQueue::flushErrorsForFile(const GlobalState &gs, FileRef file) {
+    checkOwned();
+    if (ignoreFlushes) {
+        return;
+    }
+
+    filesFlushedCount.fetch_add(1);
+    Timer timeit(tracer, "ErrorQueue::flushErrorsForFile");
+
+    core::ErrorQueueMessage msg;
+    for (auto result = queue.try_pop(msg); result.gotItem(); result = queue.try_pop(msg)) {
+        collected[msg.whatFile].emplace_back(move(msg));
+    }
+
+    vector<unique_ptr<core::ErrorQueueMessage>> errors;
+    for (auto &error : collected[file]) {
+        errors.emplace_back(make_unique<core::ErrorQueueMessage>(move(error)));
```

Comment:
qq: Can we change `collected` to be `UnorderedMap<core::FileRef, std::vector<std::unique_ptr<ErrorQueueMessage>>`? Then we can replace these loops with `move(collected[file])`.